### DISCUSSION TITLE
Can't build with CUDA OpenCL

### DISCUSSION THREAD

**Discussion by HoaSoup** [RosettaCommons/rosetta#283] [Build & Install] [answered]
Dear Developers,

I am compiling ROSETTA source code release 371 with the following command:
`./scons.py -j20 mode=release extras=opencl bin`

I stumbled this error:
`g++ -o build/src/release/linux/6.8/64/x86/gcc/13/opencl/protocols/indel/IndelOptimizationMover.os -c -std=c++0x -ffor-scope -isystem external/boost_submod/ -isystem external/ -isystem external/include/ -isystem external/dbio/ -isystem external/libxml2/include -isystem external/rdkit -pipe -Wall -Wextra -pedantic -Wno-long-long -Wno-strict-aliasing -march=core2 -mtune=generic -O3 -ffast-math -fno-finite-math-only -funroll-loops -finline-functions -finline-limit=20000 -s -Wno-unused-variable -Wno-unused-parameter -fPIC -DBOOST_ERROR_CODE_HEADER_ONLY -DBOOST_SYSTEM_NO_DEPRECATED -DBOOST_MATH_NO_LONG_DOUBLE_MATH_FUNCTIONS -DBOOST_DISABLE_THREADS -DPTR_STD -DNDEBUG -DUSEOPENCL -Isrc -Iexternal/include -Isrc/platform/linux/64/gcc/13 -Isrc/platform/linux/64/gcc -Isrc/platform/linux/64 -Isrc/platform/linux -I/opt/AMDAPP/include -I/opt/AMDAPP/include/CL -I/opt/AMDAPP/lib/x86 -I/opt/AMDAPP/lib/x86_64 src/protocols/indel/IndelOptimizationMover.cc
g++: warning: switch '-ffor-scope' is no longer supported
In file included from src/core/scoring/sc/MolecularSurfaceCalculator.hh:47,
                 from src/protocols/hotspot_hashing/SurfaceSearchPattern.cc:26:
src/basic/gpu/GPU.hh:28:10: fatal error: CL/cl_platform.h: No such file or directory
   28 | #include <CL/cl_platform.h>
      |          ^~~~~~~~~~~~~~~~~~
compilation terminated.
`

I then added the path to the CL include directory of CUDA 12.6 into the environment variable CPLUS_INCLUDE_PATH. As a test, I compiled some code that used CL/cl_platform.h. It went successfully, showing that g++ can find the header file. Yet, the compilation error of ROSETTA still persisted.

Kind regards,
HS

**Comment by vmullig** [comment, marked as answer]
Note that almost nothing in Rosetta benefits from CUDA or OpenCL.  The
simplest solution is just to do a default build.

**Reply by HoaSoup** [reply]
Thank you for your answer! Default build worked. No more issue for me.

**Comment by vmullig** [comment]
I should have added: the most multithreading support comes from
extras=cxx11thread, but even that is limited.  (Interaction graph
precomputation is efficiently multithreaded, as is peptide conformational
sampling with simple_cycpep_predict and a few other things, but that is
about it.) Mostly, we have relied either on launching many parallel
instances of Rosetta or on MPI processes for parallelism.  Efforts to make
Rosetta compatible with GPUs have largely moved on to successor projects
(Tmol and Masala).

Oh, there is limited support for a few Tensorflow models (which can be
executed on GPU) with extras=tensorflow or extras=tensorflow_gpu.
